### DISCUSSION TITLE
Keploy Proxy Fails to Connect to gRPC Server with "dial tcp 127.0.0.1:6789: connect: connection refused" Error

### DISCUSSION THREAD

**Discussion by Vibgitcode27** [keploy/keploy#2297] [Q&A] [answered]
# gRPC Keploy Integration Issue

## Problem Description

This project is encountering issues with integrating Keploy for capturing tests on a gRPC server. While the gRPC server works correctly on port `8000`, Keploy's proxy keeps attempting to connect to the incorrect port (`127.0.0.1:6789`), resulting in a connection refused error.

## Error Logs

When starting the server with Keploy, the following errors are encountered:

```bash
🐰 Keploy: 2024-09-22T22:31:17+05:30    ERROR   failed to dial the conn to destination server   {"Client IP Address": "127.0.0.1:46752", "Client ConnectionID": "0", "Destination IP Address": "127.0.0.1:6789", "Destination ConnectionID": "1", "proxy port": 16789, "server address": "127.0.0.1:6789", "error": "dial tcp 127.0.0.1:6789: connect: connection refused"}
🐰 Keploy: 2024-09-22T22:31:20+05:30    ERROR   failed to dial the conn to destination server   {"Client IP Address": "127.0.0.1:46766", "Client ConnectionID": "4", "Destination IP Address": "127.0.0.1:6789", "Destination ConnectionID": "5", "proxy port": 16789, "server address": "127.0.0.1:6789", "error": "dial tcp 127.0.0.1:6789: connect: connection refused"}
```

![Screenshot from 2024-09-22 23-08-48](https://github.com/user-attachments/assets/950d0835-c574-446d-a636-90721646ff92)

Image Link :- https://drive.google.com/file/d/1PNkN5hjydaeEwAgbNSqlnlV6qcATdj8C/view

## Current Setup

- Keploy version: 2.3.0-beta24
- gRPC server port: 8000
- gRPC framework: google.golang.org/grpc
- Proxy settings in Keploy: None explicitly configured (defaults used)

The gRPC server works perfectly when run without Keploy, but using Keploy for recording produces connection refused errors.

## Steps to Reproduce

1. Start the gRPC server using the Makefile command:
   ```bash
   make start
   ```

2. Make a client-side request:
   ```bash
   make client
   ```

3. Observe the connection refused errors in Keploy logs, where it tries to connect to 127.0.0.1:6789 instead of 127.0.0.1:8000.

## Expected Behavior

Keploy should correctly intercept gRPC traffic to the server running on port 8000 and successfully capture requests and responses.

## Project Structure

- `cmd/server/main.go`: gRPC Server implementation
- `service.go`: gRPC Service Handler
- `Makefile`: Contains commands for running the server, client, and Keploy
- `cmd/client/main.go`: gRPC Client implementation

## What Has Been Tried

- Manually ran the server and client without Keploy—works as expected.
- Tried explicitly configuring Keploy's proxy-port to 9000 as shown in the Makefile, but the issue persists.

## References

- Project Repository: [https://github.com/Vibgitcode27/keploy-grpc-prisma-go.git](https://github.com/Vibgitcode27/keploy-grpc-prisma-go.git)
- Inspiration for `cmd/server/main.go`: [Keploy grpcserver](https://github.com/keploy/keploy/blob/v0.9.1/grpc/grpcserver/server.go)

## Questions

1. Why is Keploy trying to connect to 127.0.0.1:6789 instead of 127.0.0.1:8000? Additionally, when I tried to run Keploy servers explicitly, the Keploy server showed unexpected behavior.
2. Is there any specific proxy configuration required for Keploy to correctly capture gRPC calls?
3. How can this issue be resolved to have Keploy successfully record gRPC traffic on the correct port?

**Comment by Vibgitcode27** [comment, marked as answer]
### Issue Resolution

The issue has been resolved. I was using an older version of the Go SDK, which is why it was trying to connect to a local server. I updated to the newer version using the following command:

```bash
go get -u github.com/keploy/go-sdk/v2
```
Now, the server is running properly; however, I am still unable to generate test cases. For now, I am marking this discussion as complete, but I will update this section on whether I was able to resolve this issue.

**Comment by Akshat005Chaudhary** [comment]
@Vibgitcode27 please let me know if you were able to generate test cases after successfully running the server.
